cpp/src/arrow/array/array_nested.cc

```diff
@@ -836,24 +845,32 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
 
   using OffsetArrayType = typename TypeTraits<OffsetArrowType>::ArrayType;
   const auto& typed_offsets = checked_cast<const OffsetArrayType&>(*offsets);
-  auto buffers = BufferVector({nullptr, typed_offsets.values()});
+
+  BufferVector buffers;
+  int64_t null_count;
+  if (null_bitmap != nullptr) {
+    buffers = BufferVector({std::move(null_bitmap), typed_offsets.values()});
+    null_count = null_bitmap->size();
+  } else {
+    buffers = BufferVector({null_bitmap, typed_offsets.values()});
```

Comment:
You can have `nullptr` in the place of `null_bitmap` here to make it more clear

Overall suggestion:

```cpp
BufferVector buffers;
buffers.resize(2);
int64_t null_count = 0;
if (null_bitmap) {
  null_count = ...calculate the pop-count of the buffer...
  buffers[0] = std::move(null_bitmap);
}
buffers[1] = typed_offsets.values();
```

`resize(2)` initializes buffers with two `nullptr` buffer pointers, so you only have to set when you have a `null_bitmap` and `buffers[1]` is always set.

Reply:
Done here: https://github.com/apache/arrow/pull/41956/commits/9afb881d888b8313e82dda20b83d5aa93c3be812